### PR TITLE
[flang] Set SAVE attribute for EQUIVALENCEd symbols consistently.

### PR DESCRIPTION
**Pull request by vzakhari**
Example:
```
subroutine global_sub()
  integer, dimension(4) :: iarr4=(/1,2,3,4/)
  integer, dimension(4) :: jarr4
  equivalence(iarr4,jarr4)
  call sub1
  print *, iarr4
contains
  subroutine sub1
    iarr4=jarr4((/4:1:-1/))
  end subroutine sub1
end subroutine global_sub
```

`iarr4` and `jarr4` are equivalenced via a global aggregate storage,
but the references inside `sub1` are lowered differently.
`iarr4` is accessed via the global aggregate storage, while `jarr4`
is accessed via the argument tuple. This confuses the FIR alias analysis,
that claims that a host associated entity cannot alias with a global
(if they have different source and do not have Target/Pointer
attributes deduced by the alias analysis).

I am not convinced that there is an issue in the alias analysis yet.
I think we'd better lower the accesses uniformly, i.e. if one variable
from an equivalence is lowered via the global aggregate storage, then
any other variable from this equivalence should be lowered the same way
(even if they are used via host association).

This patch makes sure that all symbols from an EQUIVALENCE get
and implicit SAVE attribute, if they do not have it already and
any symbol from the EQUIVALENCE is SAVEd (explicitly or implicitly).
This makes the further lowering consistent.

### CODE REVIEW

**Comment by jeanPerier** [comment]
> Note that in the above example iarr4 symbol is global, while
jarr4 is not. Maybe this issue could have been fixed before
the lowering by uniformly marking all symbols from the equivalence
as globals, so comments are very welcome.

I really think this is a bug in `semantics::IsSaved()` (called by `lower::symbolIsGlobal()`).
The standard is clear that `jarr4` implicitly has the SAVE attribute.

- 8.4 point 3 -> iarr4 has implicit save attribute because of the initialization
- 8.10.1.2 point 2 -> _"If any data object in an equivalence-set has the SAVE attribute, all other objects in the equivalence-set have the SAVE attribute"_ so jarr4 implicitly has the SAVE attribute because jarr4 has implicitly has it.

I would much rather a fix there, even though I cannot think of any semantic check that is bugged because of the current behavior of semantics::IsSaved().

Maybe semantics::FindEquivalenceSet could be used in semantics::IsSaved to check if any of the other symbol are Saved (with care to avoid infinite recursion I guess :) ). If you go that way, which I would recommend, please ask Peter for review too.

**Comment by vzakhari** [comment]
Thank you for the pointers, Jean!  I am going to rework it the way you suggest.

File: flang/lib/Semantics/resolve-names.cpp [hunk 1]
```diff
@@ -2390,7 +2390,7 @@ Symbol &ScopeHandler::MakeHostAssocSymbol(
   // These attributes can be redundantly reapplied without error
   // on the host-associated name, at most once (C815).
   symbol.implicitAttrs() =
-      symbol.attrs() & Attrs{Attr::ASYNCHRONOUS, Attr::VOLATILE};
+      symbol.attrs() & Attrs{Attr::ASYNCHRONOUS, Attr::VOLATILE, Attr::SAVE};
```
Comment (klausler) [inline flang/lib/Semantics/resolve-names.cpp:2393]: An explicit `SAVE` statement for a name in a nested scope must create a new symbol.  It is not possible for a `SAVE` statement in a nested scope to apply to a host-associated symbol.

**Review by klausler — CHANGES_REQUESTED** [review]

File: flang/lib/Semantics/resolve-names.cpp (hunk 1 above)
Comment (vzakhari) [reply flang/lib/Semantics/resolve-names.cpp:2393]: I had to make this change to resolve the following issue:
```
subroutine host
  integer :: x = 1
  integer :: y
  common /blk/ y
  equivalence(x,y)
contains
  subroutine inner
    x = y
  end subroutine inner
end subroutine host
```

Without propagating to the host-associated symbol the fact that `SAVE` is implicit, I would get this error:
```
./save.f90:8:5: error: The entity 'x' in COMMON block /blk/ may not have an explicit SAVE attribute
      x = y
      ^
./save.f90:8:9: error: The entity 'y' in COMMON block /blk/ may not have an explicit SAVE attribute
      x = y
          ^
bbc: semantic errors in save.f90
```

As you said, explicit `SAVE` statement in a nested scope will create a new symbol, so this code should not affect this case.

Comment (klausler) [reply flang/lib/Semantics/resolve-names.cpp:2393]: Have you tested that an explicit SAVE in the nested scope still creates a new symbol?

If it does, please at least change the comment before your change.

Comment (vzakhari) [reply flang/lib/Semantics/resolve-names.cpp:2393]: For this:
```
subroutine host
  integer :: x = 1
  integer :: y
  common /blk/ y
  equivalence(x,y)
contains
  subroutine inner
    save x
    save y
    x = y
  end subroutine inner
end subroutine host
```

I get this:
```
  Subprogram scope: host size=0 alignment=1 sourceRange=164 bytes
    .F18.0, SAVE (CompilerCreated) size=4 offset=0: ObjectEntity type: INTEGER(4) shape: 1_8:1_8 init:[INTEGER(4)::1_4]
    host (Subroutine): HostAssoc
    inner (Subroutine): Subprogram ()
    x, SAVE size=4 offset=0: ObjectEntity type: INTEGER(4) init:1_4
    y, SAVE size=4 offset=0: ObjectEntity type: INTEGER(4)
    Equivalence Sets: (x,y) (.F18.0,y)
    blk size=4 offset=0: CommonBlockDetails alignment=4: y
    Subprogram scope: inner size=8 alignment=4 sourceRange=57 bytes
      inner (Subroutine): HostAssoc
      x, SAVE (Implicit) size=4 offset=0: ObjectEntity type: REAL(4)
      y, SAVE (Implicit) size=4 offset=4: ObjectEntity type: REAL(4)
```

So the new symbols are created.

I will upload an update shortly.

**Comment by vzakhari** [comment]
It looks like the code formatting check is broken for some other PRs, so it should not be my problem.

More extensive testing uncovered the following problem.

```
module equiv
  real dx,dy
  common /blk/ dx
  equivalence(dx,dy)
end module
```

With my changes I get the following module file:
```
module equiv
real(4),save::dx
real(4),save::dy
common/blk/dx
equivalence(dx,dy)
end
```

So the `SAVE` was explicitly printed, which causes `error: The entity 'dx' in COMMON block /blk/ may not have an explicit SAVE attribute` when the module is used.

I am planning to filter out the implicit `SAVE` in `mod-file.cpp`, but please let me know if I am on a wrong path.

**Comment by klausler** [comment]
> It looks like the code formatting check is broken for some other PRs, so it should not be my problem.
> 
> More extensive testing uncovered the following problem.
> 
> ```
> module equiv
>   real dx,dy
>   common /blk/ dx
>   equivalence(dx,dy)
> end module
> ```
> 
> With my changes I get the following module file:
> 
> ```
> module equiv
> real(4),save::dx
> real(4),save::dy
> common/blk/dx
> equivalence(dx,dy)
> end
> ```
> 
> So the `SAVE` was explicitly printed, which causes `error: The entity 'dx' in COMMON block /blk/ may not have an explicit SAVE attribute` when the module is used.
> 
> I am planning to filter out the implicit `SAVE` in `mod-file.cpp`, but please let me know if I am on a wrong path.

The `SAVE` attribute should not have been emitted to the module file if it was only an implicit attribute, as it should have been in this case.  (If it is not an implicit attribute in the symbol table, that's the problem instead.).

**Review by klausler — APPROVED** [review]

**Comment by vzakhari** [comment]
Thanks! Yes, it is implicit in the symbol table, so I am using this info to filter it out.  Only 2 tests are affected by this, so I will fix it in a follow-up PR.

**Review by jeanPerier — APPROVED** [review]
Thanks for the move of the fix into semantics Slava!

**Pull request merged**